paddle/fluid/framework/details/multi_devices_graph_pass.cc

```diff
@@ -862,29 +861,118 @@ int MultiDevSSAGraphBuilder::CreateRPCOp(
+        "the distribute training related op should be in [split_byref, "
+        "concat].");
+  }
+
+  PADDLE_ENFORCE(op_dev_id != -1,
+                 "can not find right place for distributed op: %s",
+                 node->Op()->Type());
+
+  CreateComputationalOp(result, node, op_dev_id);
+  return op_dev_id;
 }
 
-bool MultiDevSSAGraphBuilder::IsScaleLossOp(ir::Node *node) const {
-  return boost::get<int>(
-             node->Op()->GetAttr(OpProtoAndCheckerMaker::OpRoleAttrName())) ==
-             (static_cast<int>(OpRole::kBackward) |
-              static_cast<int>(OpRole::kLoss)) &&
-         !loss_var_name_.empty();  // If loss_var is empty. This is test mode
+void DistSSAGraphBuilder::CreateCollectionOp(ir::Graph *result,
+                                             const std::string &p_name,
+                                             const std::string &g_name) const {
+  size_t cur_device_id = 0;
+  switch (strategy_.reduce_) {
+    case BuildStrategy::ReduceStrategy::kReduce:
+      cur_device_id = GetAppropriateDeviceID({g_name});
+      CreateReduceOp(result, g_name, cur_device_id);
+      sharded_var_device_.emplace(g_name, cur_device_id);
+      break;
+    case BuildStrategy::ReduceStrategy::kAllReduce:
+      if (IsSparseGradient(g_name)) {
+        CreateReduceOp(result, g_name, 0);
+        CreateBroadcastOp(result, g_name, 0);
+      } else {
+        InsertAllReduceOp(result, g_name);
+      }
+      break;
+    default:
+      LOG(FATAL) << "Unknown reduce strategy.";
+      break;
+  }
+}
+
+void DistSSAGraphBuilder::InsertPostprocessOps(ir::Graph *result) const {
+  if (strategy_.fuse_broadcast_op_) {
+    CreateFusedBroadcastOp(result, bcast_var_name_set_);
+  } else {
+    for (size_t dev_id = 0; dev_id < bcast_var_name_set_.size(); ++dev_id) {
+      auto &to_bcast_set = bcast_var_name_set_[dev_id];
+      for (auto &bcast_name : to_bcast_set) {
+        CreateBroadcastOp(result, bcast_name, dev_id);
+      }
+    }
+  }
 }
+
+bool DistSSAGraphBuilder::NeedCollectionOps() const { return true; }
+
+std::unordered_set<std::string> &MultiDevSSAGraphBuilderRegistry() {
+  static std::unordered_set<std::string> regs;
+  return regs;
+}
+
+int RegisterMultiDevSSAGraphBuilder(const std::string &builder_mode) {
+  MultiDevSSAGraphBuilderRegistry().insert(builder_mode);
+  return 0;
+}
+
 }  // namespace details
 }  // namespace framework
 }  // namespace paddle
 
-REGISTER_PASS(multi_devices_pass,
-              paddle::framework::details::MultiDevSSAGraphBuilder)
-    .RequirePassAttr(paddle::framework::details::kLossVarName)
-    .RequirePassAttr(paddle::framework::details::kPlaces)
-    .RequirePassAttr(paddle::framework::details::kLocalScopes)
-    .RequirePassAttr(paddle::framework::details::kStrategy)
-    .RequirePassAttr(paddle::framework::details::kNumTrainers);
+#define REGISTER_MULTI_DEVICES_PASS(pass_name, pass_class)       \
+  REGISTER_MULTI_DEV_PASS(pass_name);                            \
```

Comment:
What's the difference between `REGISTER_MULTI_DEV_PASS ` and `REGISTER_MULTI_DEVICES_PASS ` ?

Reply:
`REGISTER_MULTI_DEV_PASS ` has been removed.